Improving LLVM tests

nikolaperic opened the pull request:
These changes fix the remaining 13 failing LLVM tests when running `ninja check-llvm`:
- CodeGen/Mips/nanomips/bittestbranch.ll
- CodeGen/Mips/nanomips/jumptable.ll
- MC/Mips/cnmips/invalid.s
- MC/Mips/macro-li-bad.s
- MC/Mips/mips32r6/invalid.s
- MC/Mips/micromips-invalid.s
- MC/Mips/mips32r2/invalid.s
- MC/Mips/micromips32r6/invalid.s
- MC/Mips/mips64r6/invalid.s
- MC/Mips/msa/invalid-64.s
- MC/Mips/msa/invalid.s
- tools/UpdateTestChecks/update_llc_test_checks/mips_generated_funcs.test
- DebugInfo/Mips/delay-slot.ll

Commit:
Improving LLVM tests

- Fix instructions conflicts between NanoMips and other Mips targets
  that caused tests from the MC/Mips group to fail
- Fix several tests
- Fix debug location of SP adjustment instruction that caused
  DebugInfo/Mips/delay-slot.ll test to fail
- Add 'nanomips' feature to llvm-lit config script so that
  tests can be marked as UNSUPPORTED for NanoMips target

farazs-github commented on llvm/utils/lit/lit/llvm/config.py:
```diff
@@ -121,7 +121,9 @@ def __init__(self, lit_config, config):
             elif re.match(r'^arm.*', target_triple):
                 features.add('target-arm')
-
+        
```
trailing whitespace

farazs-github commented on llvm/lib/Target/Mips/NanoMipsInstrInfo.td:
```diff
@@ -358,6 +358,11 @@ def XORI_NM  : ArithLogicI32<"xori", uimm12_nm, GPR32NMOpnd, imm32ZExt12, xor>;
 def SLTI_NM  : ArithLogicI32<"slti", uimm12_nm, GPR32NMOpnd, imm32ZExt12, setlt>;
 def SLTIU_NM : ArithLogicI32<"sltiu", uimm12_nm, GPR32NMOpnd, imm32ZExt12, setult>;
+
+// TODO: isCodeGenOnly=1 was added to avoid conflicts with same instructions 
```
cleanup trailing whitespaces in comments ... in this and other places

farazs-github commented on llvm/test/tools/UpdateTestChecks/update_llc_test_checks/mips_generated_funcs.test:
```diff
@@ -1,3 +1,6 @@
+# UNSUPPORTED: nanomips
```
Does this test fail only for nanoMIPS? I would imagine that if outliner is a problem for this test and outliner is enabled for all MIPS, then the test should fail for all MIPS.

nikolaperic replied:
On the second look I noticed that the test was written with the expectation of using the outliner, because we have `-enable-machine-outliner` option for `llc` tool inside `mips_generated_funcs.ll` file. Inside that file there is also a NOTE comment suggesting that outlining was not working at the time the test was written. I assume that this is why the expected output is written without outlined parts of the code. I think the expected output in files `mips_generated_funcs.ll.generated.expected` and `mips_generated_funcs.ll.nogenerated.expected` should be changed to expect outlined parts of the code since outliner is now working. In that case test will not fail. 

What do you think about that?

farazs-github replied:
I agree, fixing the expected output looks like the right solution.

farazs-github commented on llvm/test/CodeGen/Mips/nanomips/jumptable.ll:
```diff
@@ -8,11 +8,14 @@ define i32 @test_jumptable(i32 %in) {
-; CHECK: jrc $a0
+; CHECK: la $a1
+; CHECK: LA_NM
+; CHECK: .reloc .Ltmp0, R_NANOMIPS_JUMPTABLE_LOAD, .LJTI0_0
```
IMO fixing this test should be limited to making it work for the case it was written for - that is regular jump tables. Compressed tables should be checked by a separate test if they aren't already.

nikolaperic replied:
Thanks for the comment, I will fix that.

farazs-github commented on llvm/lib/Target/Mips/MipsSEInstrInfo.cpp:
```diff
@@ -625,7 +625,9 @@ void MipsSEInstrInfo::adjustStackPtr(unsigned SP, int64_t Amount,
   MipsABIInfo ABI = Subtarget.getABI();
   DebugLoc DL;
-  if (I != MBB.end()) DL = I->getDebugLoc();
+  // NOTE: If DL is set this way test DebugInfo/Mips/delay-slot.ll will fail
```
The fact that a test is failing is not sufficient reason to change code. I'd like to understand why this line was added in the first place and what effect removing it will have.

@cme might have some input here

cme replied:
Yes, this was added in response to failing tests in the GDB test suite where the user debug experience was improved by having a location set on the stack increment. I think without it, the debugger would see the adjusted stack value, but still have the same in-function call frame information when stopped at a 'return' line, which is misleading for the user. I'll check whether the debug information in the test is actually more likely to be 'correct' with or without. 

However, since this test is specifically a *delay slot* test, it seems likely that this is somewhere that the reasoning about "the right thing" may be divergent between NanoMips and classic MIPS, so making the debug line setting conditional on `.haNanoMips()` might be the right thing to do.

nikolaperic replied:
Thanks for the detailed information. I left that part of the code under comment because I assumed it has some purpose. It makes sense now to condition it with `.hasNanoMips()`. I will change that.

farazs-github reviewed (changes requested):
Comments added. At least one change needs further inputs from @cme

farazs-github commented on llvm/test/CodeGen/Mips/nanomips/jumptable.ll:
```diff
@@ -8,11 +8,11 @@ define i32 @test_jumptable(i32 %in) {
-; CHECK: jrc $a0
+; CHECK: la $a1
+; CHECK: LA_NM
+; CHECK: lbux $a0, $a0($a1)
```
Guess I was not explicit enough. This test should continue to check the LWXS expansion of jump tables, by being compiled with `-mattr=disable-jump-table-opt`. Changing li -> la is correct, but changing lwxs -> lbux is contrary to the intention of this test.

nikolaperic replied:
Okay, I understand now, I'll fix it. Thank you.

farazs-github reviewed (changes requested):
Few more changes required.

I suggest you add more commits instead of force-pushing on the pull request. I'll squash-and-merge all the changes once approved to get a clean history.

Commit:
Further LLVM tests improvement

- Fix expected output for jumptable.ll and
  mips_generated_funcs.ll tests

Commit:
Fixing failing clang tests

- No longer need to mark a test ExtDebugInfo.cpp that
  needs object emitter as UNSUPPORTED
- Delete line left behind by a merge conflict in
  sanitize-coverage-old-pm.c test

nikolaperic commented:
Added one more commit which fixes last two of failing clang tests. 

Test `clang/test/CodeGen/sanitize-coverage-old-pm.c` now passes unexpectedly because it was marked as `XFAIL`. Was that the intention or maybe the test should have been marked as `UNSUPPORTED`? It was marked as `XFAIL` in [this commit](https://github.com/MediaTek-Labs/llvm-project/commit/4d8bbe8eadd6ab3613988ac6a671287987dd919c).

farazs-github commented on clang/test/CodeGen/sanitize-coverage-old-pm.c:
```diff
@@ -9,7 +9,6 @@
 // UNSUPPORTED: armv7, armv7l, thumbv7, armv8l
 // The same issue also occurs on a riscv32 host.
 // XFAIL: riscv32, nanomips
```
I think UNSUPPORTED is the correct choice here.
XFAIL is for documenting known bugs or issues.

farazs-github reviewed (changes requested):
Minor change

farazs-github reviewed (approved):
LGTM

Commit:
Mark sanitize-coverage-old-pm.c as unsupported
